### DISCUSSION TITLE
Request needs to send json field, but data is the default

### DISCUSSION THREAD

**Discussion by ElSnoMan** [ElSnoMan/pyclinic#38] [Q&A]
Why would my test_create_todo() not be working here?  The test_get_todos() works.  When I just build it out manually with requests it works.

```python
import pytest
from pyclinic.postman import Postman


@pytest.fixture
def todo_runner():
    return Postman("Todo-Flask-Heroku.postman_collection.json")
```

```python
import requests
from pyclinic.postman import Postman
from requests.models import Response


def test_get_todos(todo_runner):
    response = todo_runner.Root.get_todos()
    assert response.status_code == 200


def test_create_todo(todo_runner):
    response = todo_runner.Root.post_todo()
    assert response.status_code == 200


def test_create_todo_with_requests():
    data = {"title": "hello from request", "done": False}
    response = requests.post(
        "https://df-flask-todo-api-class.herokuapp.com/todo", json=data
    )
    assert response.status_code == 200
```

I also tried it like this:

```python
def test_create_todo(todo_runner):
    # response = todo_runner.Root.post_todo()
    response = todo_runner.Root.post_todo(
        data={"title": "hello from request", "done": False}
    )
    assert response.status_code == 200
```

[Todo-Flask-Heroku.postman_collection.json](https://app.gitkraken.com/api/glo/boards/60ff0e574119dd0012211b17/attachments/61169e49e979b40012bddbfb)

**Comment by ElSnoMan** [comment]
@gleekzorp

When I inspected the request with `runner.Root.post_todo.help()`, I saw that the headers had the Content-Type set to `application/json`. Because of this, you need to replace the _default_ `data` field with a `json` field:

```python
runner = Postman(path)
runner.Root.post_todo.help()  # 1

req = runner.Root.post_todo.request
req["json"] = req.pop("data")

runner.Root.post_todo.help() # 2
response = runner.Root.post_todo()
assert response.ok
```

The request dictionary to send at `#1`
```javascript
{
    'method': 'POST',
    'url': 'https://df-flask-todo-api-class.herokuapp.com/todo',
    'data': {'title': 'check new stuff', 'done': False},
    'headers': {'Content-Type': 'application/json'}
}
```

After renaming `data` to `json`, here's what the request looks like in `#2`
```javascript
{
    'method': 'POST',
    'url': 'https://df-flask-todo-api-class.herokuapp.com/todo',
    'headers': {'Content-Type': 'application/json'},
    'json': {'title': 'check new stuff', 'done': False}
}
```

**Reply by gleekzorp** [reply]
Thank you for the help.  I might be a little confused though.  Would I have to always do this override?  Below is a screenshot of what my postman collection looks like.  Did I setup my postman collection wrong?

![todo-postman](https://user-images.githubusercontent.com/34635302/129401885-1473e692-cdc5-4c71-a803-cc3852f250ad.png)

Will this be how I would fix this?  It isn't working for me.

```python
def test_create_todo(todo_runner):
    req = todo_runner.Root.post_todo.request
    req["json"] = req.pop("data")
    response = todo_runner.Root.post_todo()
    assert response.status_code == 200
```

**Reply by gleekzorp** [reply]
Uh... Maybe I didn't save my file or something, it is working now.

**Reply by gleekzorp** [reply]
Please take a look at the image below it has the headers for that request.  As you can see I have two `Content-Type: application/json`.  When you hover the first content-type `i` it says this header was automatically added, so obviously that would be based on the body that I'm sending, and the second one I somehow added manually since the second one doesn't have the `i`.
![headers](https://user-images.githubusercontent.com/34635302/129407803-efacac73-90a7-4446-9e6f-1a19c409077c.png)

I completely remade that request and exported my postman collection.  Now when I call `runner.Root.post_todo.help()` I see that my headers are empty.
```
{
    'method': 'POST',
    'url': 'https://df-flask-todo-api-class.herokuapp.com/todo',
    'data': {'title': 'check new stuff', 'done': False},
    'headers': {}
}
```

When I go back to the following test it will fail again
```python
def test_create_todo(todo_runner):
    response = todo_runner.Root.post_todo()
    assert response.status_code == 200
```

I looked at the logs on my API and I get the following error.  I guess this could be my bad on how I setup my API but the API only accepts `json` data.

```
title = request.json['title']
TypeError: 'NoneType' object is not subscriptable
```

The API was created with `flask` and `flask-sqlalchemy`, here is a little snippet of that route.
```python
@app.route("/todo", methods=["POST"])
def add_todo():
  title = request.json["title"]
  done = request.json["done"]

  new_todo = Todo(title, done)

  db.session.add(new_todo)
  db.session.commit()

  todo = Todo.query.get(new_todo.id)
  return todo_schema.jsonify(todo)
```

Throwing an idea out there.  In your `build_request_to_send` function you have a single check for if the request body is raw, but it seems there are multiple `options` for raw such as `text`, `javascript`, `json`, `html`, and `xml`.  Should we have some more checks here to possibly set up the request differently based on that raw data?

In this function, you add a `data` key if it is raw but if it is `raw/json` I would think you would need to add a `json` key.  See the requests docs for why my thought process is going this way https://docs.python-requests.org/en/latest/user/quickstart/#more-complicated-post-requests

They show two ways of handling this json data:
```python
>>> import json

>>> url = 'https://api.github.com/some/endpoint'
>>> payload = {'some': 'data'}

>>> r = requests.post(url, data=json.dumps(payload))
```
```python
>>> url = 'https://api.github.com/some/endpoint'
>>> payload = {'some': 'data'}

>>> r = requests.post(url, json=payload)
```

Here is my request and as you can see there is an `options` key.
```
{
  "name": "Post Todo",
  "request": {
    "method": "POST",
    "header": [],
    "body": {
      "mode": "raw",
      "raw": "{\n\t\"title\": \"check new stuff\",\n\t\"done\": false\n}",
      "options": {
        "raw": {
          "language": "json"
        }
      }
    },
    "url": {
      "raw": "https://df-flask-todo-api-class.herokuapp.com/todo",
      "protocol": "https",
      "host": [
        "df-flask-todo-api-class",
        "herokuapp",
        "com"
      ],
      "path": [
        "todo"
      ]
    }
  }
}
```

Here is your `build_request_to_send` located in `pyclinic/postman.py`
```python
def build_request_to_send(request: PostmanRequest, variables: Dict) -> Dict:
    """Convert a single PostmanRequest to a dictionary to be used by requests."""
    url = replace_url_variables_with_actual_values(request, variables)

    # Build the request dictionary (method, url, headers, data, etc.)
    request_dict = request.dict()
    if request.body:
        if request.body.get("raw"):
            request_dict["data"] = json.loads(request.body.get("raw"))
        elif request.body.get("urlencoded"):
            request_dict["data"] = str(request.body.get("urlencoded"))
        else:  # TODO: handle other types of request body like GraphQL and form-data
            pass
    del request_dict["body"]  # uses ["data"] instead of ["body"]
    request_dict["url"] = url

    request_dict["headers"] = {}
    header = request_dict.pop("header")
    if header:
        for item in header:
            key = item["key"]
            value = item["value"]
            request_dict["headers"].update({key: value})
    return request_dict
```

I hope I'm on the right track here.  If I run into some time this weekend I will fork your code and see if my idea pans out.

**Reply by ElSnoMan** [reply]
We can try something like the `json.dumps()` and see if that makes it better, but do know that the examples we're using also has that `options` field and it's been working so far. He's an example of the `IsAuthorized` request from the `bookstore.postman_collection.json`

```javascript
"request": {
    "method": "POST",
    "header": [],
    "body": {
        "mode": "raw",
        "raw": "{\n  \"userName\": \"Carlos Kidman\",\n  \"password\": \"P@$$w0rd\"\n}",
        "options": {
            "raw": {
                "language": "json"
            }
        }
    },
    "url": {
        "raw": "{{BASE_URL}}/Account/v1/Authorized",
        "host": [
            "{{BASE_URL}}"
        ],
        "path": [
            "Account",
	    "v1",
	    "Authorized"
        ]
    }
}
```

Your endpoint is looking specifically for the `json` field to be populated, so it seems more like an implementation detail that has a workaround. If we changed it do send `json` instead, but your endpoint looked for `request.data`, my hypothesis would be that you would run into the same problem.

Since PyClinic doesn't know about your implementation details, we need to find a way to make it so then it's generalized enough to handle when an endpoint requires the `data` field and when it needs the `json` field.

🎉  Locally, I changed the `build_request_to_send` function to use `json.dumps()` instead and all tests passed. Hopefully `json.dumps()` is able to solve that problem for you too.

P.S. Maybe another solution on the endpoint's side would be to check the json and data fields and use the one that isn't empty. That's _another_ idea haha. So many ways we can go about this!

```python
@app.route("/todo", methods=["POST"])
def add_todo():
    data = request.json if request.json else request.data
    title = data["title"]
    done = data["done"]

    new_todo = Todo(title, done)

    db.session.add(new_todo)
    db.session.commit()

    todo = Todo.query.get(new_todo.id)
    return todo_schema.jsonify(todo)
```
